src/transformers/models/bloom/convert_bloom_original_checkpoint_to_pytorch.py

```diff
@@ -0,0 +1,262 @@
+                    temp[layer_name_mapping(key, file)] = temp.pop(key)
+
+                if tensors is None:
+                    tensors = temp
+                else:
+                    for key in tensors.keys():
+                        if any(key.endswith(end) for end in WEIGHTS_TO_AVERAGE_ENDSWITH):
+                            tensors[key] += temp[
+                                key
+                            ]  # We average (sum and then divide) some weights accross TP ranks (see https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/olruwase/sync_layer_norms/megatron/training.py#L425)
+                        else:
+                            cat_dim = (
+                                1 if any(text in key for text in WEIGHTS_WITH_ROW_PARALLELISM_CONTAIN) else 0
+                            )  # Some weights are RowParallelLinear in Megatron-Deepspeed, others are ColumnParallel
+                            tensors[key] = torch.cat(
+                                [tensors[key], temp[key]], dim=cat_dim
+                            )  # We concatenate these weights accross TP ranks
+
+            # Divide by the number of TP the weights we want to average
+            for key in tensors.keys():
+                if any(key.endswith(end) for end in WEIGHTS_TO_AVERAGE_ENDSWITH):
+                    tensors[key] = tensors[key] / config.pretraining_tp
+                # final_tensors['transformer.'+key] = tensors[key]
+            torch.save(
+                tensors,
+                os.path.join(
+                    pytorch_dump_folder_path,
+                    "pytorch_model_{}-of-{}.bin".format(str(j + 1).zfill(5), str(len(file_names)).zfill(5)),
+                ),
+            )
+
+            tensor_size = 1
+            for key in tensors.keys():
+                temp_tensor = tensors[key]
+                for ele in temp_tensor.size():
+                    tensor_size *= ele
+                total_size += dtype_size * tensor_size
+                if any(key.endswith(end) for end in WEIGHTS_TO_AVERAGE_ENDSWITH):
+                    tensors[key] = tensors[key] / config.pretraining_tp
+
+                if key not in index_dict["weight_map"]:
+                    index_dict["weight_map"][key] = "pytorch_model_{}-of-{}.bin".format(
+                        str(j + 1).zfill(5), str(len(file_names)).zfill(5)
+                    )
+
+        config = BLOOMConfig()
+        pytorch_config_dump_path = pytorch_dump_folder_path + "/" + CONFIG_NAME
+        index_dict["metadata"]["total_size"] = total_size
+        with open(pytorch_config_dump_path, "w", encoding="utf-8") as f:
+            f.write(config.to_json_string())
+        with open(os.path.join(pytorch_dump_folder_path, WEIGHTS_NAME + ".index.json"), "w", encoding="utf-8") as f:
+            json_config = json.dumps(index_dict, indent=2, sort_keys=True) + "\n"
+            f.write(json_config)
+    else:
+        model = BLOOMModel(config)
```

Comment:
why do you need to create a model if you are not really using it?

But the main reason for not creating it is that it requires much much more memory, especially since you're creating it in fp32 so this will require an additional 176*4=704GB of CPU RAM.

Did you actually try it on the 176B size?

Also remember that all those temporary vars when you `cat` tensors all consume additional memory. So it'll add up very quickly.

I wonder if we should just call into `modeling_utils' final model sharding mechanism directly and never need to create the full model in the conversion script.

(I'm just brainstorming - we can discuss on slack - might be easier)

Reply:
For reference of feasibility - the A100s on JZ are 512GB and our dgx too is 512GB CPU RAM.

Reply:
Thanks for the comment !
Indeed, but the model is used to make sure that the keys match and we are not missing any key. When I dealt with larger models I use the sharded version of the script which does not instantiate the model when starting the process. What I propose is to create the model as it is done in accelerate `init_empty_weights`

Reply:
Hmm actually it is a bit tricky to use `init_empty_weights` from accelerate. Even with that we still need to load dynamically the weights to check if there are any unexpected keys. 
I think that I would either remove the whole process, or adding an if statement that checks for eg the number of layers + hidden dimension and advice the user to use the sharded version of the conversion